feat(website): Adapt collectionDate search to work with date ranges

fhennig opened the pull request:
<!-- Mention the issue that this PR resolves. Delete if there is no corresponding issue -->
resolves #3285 

<!-- Add "preview" label in almost all cases to have testable deployment. Then lookup the deployment URL in argocd and paste it here (if you don't know how to look up the URL, ask here: https://loculus.slack.com/archives/C06JCAZLG14), it's something like `{REPLACE}.loculus.org` -->
preview URL: https://date-ranges-ui.loculus.org

### Summary
<!-- Add a few sentences describing the main changes introduced in this PR -->
Add config bit to allow configuring the `rangeOverlapSearch` on a property (or rather two properties belonging together). Add a new `DateRangeField` search field type for these kinds of date ranges.


This is the first time we don't have a 1:1 mapping of search fields and API parameters anymore, so some bits in the code are a bit ... complicated maybe? But so far this way of doing it still seems fine. A "proper" solution would (IMO) involve decoupling the metadata field definition and the search field definitions, but seems kind of heavy handed at the moment.


### Screenshot
I wanted to maybe use a toggle, but ran into [this UI bug](https://github.com/saadeghi/daisyui/discussions/2451) (clash with tailwind forms).

<img width="303" alt="image" src="https://github.com/user-attachments/assets/2da3e712-50d5-4e05-9253-cef42599bd7f">

#### with tooltip

<img width="523" alt="image" src="https://github.com/user-attachments/assets/9b11fd3b-3056-4ed8-98d9-e0bfd5360d37">

### PR Checklist
<!-- Check completed items of strikethrough irrelevant items (using ~~text~~) -->
- [x] All necessary documentation has been adapted.
- [x] The implemented feature is covered by an appropriate test.

fhennig commented on website/src/utils/search.ts:
```diff
@@ -63,7 +63,13 @@ const getFieldOrColumnVisibilitiesFromQuery = (
         }
-        visibilities.set(field.name, initiallyVisibleAccessor(field) === true);
+
+        let fieldName = field.name;
+
+        if (field.rangeOverlapSearch) {
+            fieldName = field.rangeOverlapSearch.rangeName;
+        }
+        visibilities.set(fieldName, initiallyVisibleAccessor(field) === true);
```
The modifications in this function could be avoided if we force users to uses a `field.name` as the `rangeName`. This can't really be enforced. IMO this can't be avoided because we have to group fields two distinct fields.

Happy for input though if anybody happens to look at the PR already!

theosanderson replied:
I think we need some more changes in this general area (appreciate the PR is still a draft): atm Collection date shows up twice on the list of fields to change the visibility of. I'd need to think about exactly the best way to handle this.

fhennig replied:
I think it's just there twice because of the second _Collection date_ `string` field which is just above the range field.

I had a vague feeling that we shouldn't look for Metadata field visiblity, but rather MetadataFilter visibility! But also haven't looked into it more yet. But good to know that you think this area could use some more work! I'll spend some more time on it then (probably monday).

theosanderson replied:
> I think it's just there twice because of the second Collection date string field which is just above the range field.

OK! That may well be true - my bad - and in that case it may well not need more work - sorry for my misunderstanding!

fhennig commented on website/src/utils/search.ts:
```diff
@@ -63,7 +63,13 @@ const getFieldOrColumnVisibilitiesFromQuery = (
         }
-        visibilities.set(field.name, initiallyVisibleAccessor(field) === true);
+
+        let fieldName = field.name;
+
+        if (field.rangeOverlapSearch) {
+            fieldName = field.rangeOverlapSearch.rangeName;
+        }
+        visibilities.set(fieldName, initiallyVisibleAccessor(field) === true);
```
The modifications in this function could be avoided if we force users to uses a `field.name` as the `rangeName`. This can't really be enforced. IMO this can't be avoided because we have to group fields two distinct fields.

Happy for input though if anybody happens to look at the PR already!

theosanderson commented:
The preview doesn't seem to be showing this UI for me atm?

fhennig commented:
> The preview doesn't seem to be showing this UI for me atm?

Yes I noticed as well, was a bit confused, it was due to the rangeSearch being enabled because of the `date` type and it taking precedence in the code. I think The last commit will fix it.

theosanderson commented on website/src/components/SearchPage/fields/DateRangeField.tsx:
```diff
@@ -0,0 +1,134 @@
+            <DateField
+                field={{
+                    name: 'not used',
+                    label: 'Lower',
```
IMO `From` and `To` is significantly more understandable

theosanderson replied:
(to be clear, for code the current is fine, and arguably better given we use `From` for the individual LAPIS query - just talking in terms of UI)

fhennig replied:
https://github.com/loculus-project/loculus/pull/3308/commits/1daa04f01bd7f9b8f2db084d6e18a4204a85accf

theosanderson commented:
(looks really nice!  will try to dive in more soon!)

theosanderson commented on website/src/components/SearchPage/fields/DateRangeField.tsx:
```diff
@@ -0,0 +1,134 @@
+                    <span className='text-gray-400 text-sm mr-2'>strict</span>
+                    <input
+                        type='checkbox'
+                        className='checkbox checkbox-sm [--chkbg:theme(colors.gray.400)]'
```
> the checkbox looks a bit like it's disabled, but with more solid colors it draws a lot of attention. But making it smaller would also not be great for clicking it I think. 🤔 Haven't come up with a good solution for this yet. (but maybe not critical, I don't think it looks terrible)

I don't think it looks terrible and I don't think we need to block, but I imagine it could be better with white bg and gray tick rather than the reverse

fhennig replied:
That's a great idea! I implemented it in https://github.com/loculus-project/loculus/pull/3308/commits/5d5ea4f354871414fc4b513fd1f29957e46c6e2c (also updated the screenshots)

fhennig commented:
Next I think I want to add a test for new new component

corneliusroemer commented:
Looks great! 

For test, did you think of a component test? For E2E we'd need to modify test data, which might be possible as it's only an addition we need here, but we could also wait for Theo's more realistic new test setup.

corneliusroemer commented:
@fhennig sorry for force pushing the base branch, this causes merge conflicts, I'll try to resolve them - note that the last two pushes had no CI running due to the conflicts: 

<img width="947" alt="image" src="https://github.com/user-attachments/assets/0785eb3d-3f8d-4026-8b8d-5dde02a262fc">

corneliusroemer commented:
The rebase was easy, but there was another conflict due to reordering of values.yaml, I hope I did it correctly, if there are issues beware of this potentially being the source @fhennig

fhennig commented on kubernetes/loculus/values.yaml:
```diff
@@ -101,7 +91,24 @@ defaultOrganismConfig: &defaultOrganismConfig
+          bound: upper
         noInput: true
+      - name: sampleCollectionDate
+        displayName: Collection date (text)
```
Is the displayName also used as the column title?

theosanderson replied:
So we can add `notSearchable: true` for search.

In terms of columns it looks like we have a bug here in that there is a `Collection date` in the column selector which when ticked doesn't make a column appear. (For understandable reasons!)

fhennig replied:
> In terms of columns it looks like we have a bug here in that there is a Collection date in the column selector which when ticked doesn't make a column appear

Should we just tolerate this and make an issue for it and try and adress soonish? Besides this I think the solution is good.

fhennig replied:
I removed the "(text)" and added the "notSearchable" bit

theosanderson replied:
(i.e. we can remove it specifically by name/id or whatever in a totally hardcoded way)

fhennig replied:
hacky fix: https://github.com/loculus-project/loculus/pull/3308/commits/4cc6f4a59f225b7aaf55e3a406900d43b543805d

issue: https://github.com/loculus-project/loculus/issues/3325

`notSearchable` didn't actually work, I fixed it here: https://github.com/loculus-project/loculus/pull/3308/commits/2cca86efecc3cba09fb0e0dffa64ea00ca1ee00c - not sure if this might have other consequences we didn't yet think about?

corneliusroemer commented:
Did another merge conflict resolution post-merge of the base, this should have been the last resolution we need.

corneliusroemer reviewed (changes requested):
I _think_ the current implementation doesn't allow going from URL to search state.

When I uncheck the strict box and have a date in the bar, copy to new browser tab, I get a different result.

It seems that the strictness checkbox deselection state isn't correctly deduced from the presence of `UpperFrom` (which indicates lax search) https://date-ranges-ui.loculus.org/west-nile/search?sampleCollectionDateRangeUpperFrom=2024-07-09

Instead of checkbox deselected and date 2024-07-09 being in "from" part of the date range picker, I get strict being selected and no date being entered:

<img width="633" alt="image" src="https://github.com/user-attachments/assets/a70c610e-6b06-4bac-8429-d7ace0d92cb6">

fhennig commented:
@corneliusroemer Ah thanks for checking that! I'll fix it. I somehow completely forgot that you don't need to always give both sides of the range. If you give both sides of the range, it should correclty infer it from the URL. I know where to make the changes, I'll go ahead and do so!

theosanderson reviewed (approved):
LGTM - really great work, thanks

corneliusroemer commented:
This still doesn't work: https://date-ranges-ui.loculus.org/west-nile/search?sampleCollectionDateRangeUpperFrom=2024-10-14 - the URL params get wiped and UI is without

fhennig commented:
Ah yes I see it too for the not-strict case. I'll get to it tomorrow

chaoran-chen commented on kubernetes/loculus/templates/_common-metadata.tpl:
```diff
@@ -202,6 +202,9 @@ organisms:
   {{- if or (or (eq .type "timestamp") (eq .type "date")) .rangeSearch }}
   rangeSearch: true
   {{- end }}
+  {{- if .rangeOverlapSearch }}
+  rangeOverlapSearch: {{ .rangeOverlapSearch | toJson }}
+  {{- end}}
```
Could you please document the new config options in `/docs`?

chaoran-chen commented on website/src/components/SearchPage/fields/DateRangeField.tsx:
```diff
@@ -0,0 +1,131 @@
+                            inside of the search range.
+                        </p>
+                        <p>
+                            <span className='font-bold'>not strict: </span>
+                            {field.displayName} range must have <span className='italic'>some overlap </span>
+                            with the search range.
```
I think that it would be great to have some description about this in the [search sequences documentation](https://loculus.org/for-users/search-sequences/). The tooltips are already really nice but on the documentation page, we can a bit more precise and maybe have an example?

fhennig replied:
https://docs-date-ranges-ui.loculus.org/for-users/search-sequences/

(Image not loading yet, maybe just not deployed yet?)

chaoran-chen replied:
Nice! Looks amazing!

chaoran-chen reviewed (approved):
Thanks very much, Felix! I didn't look at the code but only tested it and it looks really nice! I just have two reminders to add some documentation.

fhennig commented:
Wow getting my image to work in starlight is a pain. It won't take the SVG and it's scaling the PNG to look blurry.

anna-parker commented:
@fhennig not sure what image you mean but https://pictogrammers.com/library/mdi/ has a ton of great images that can be easily resized and are free for use.

fhennig commented:
@anna-parker Ah I meant the diagram i added to the docs

corneliusroemer commented:
Thanks for fixing the bug, indeed it works now!

Something I noticed that doesn't need fixing: with lax mode we show results even if the range is negative with starting date _after_ ending date. That's because we use 2 dates here rather than a single range picker - which is good because a range picker would always require start+end when we often just want one of the two

<img width="655" alt="Slack 2024-12-03 17 15 43" src="https://github.com/user-attachments/assets/f7990ffd-5544-4290-8045-05bf492f3718">

corneliusroemer commented on website/src/types/config.ts:
```diff
@@ -27,6 +27,19 @@ export const customDisplay = z.object({
+ */
+export const rangeOverlapSearch = z.object({
+    /**
+     * specify rangeName in both (upper and lower) fields to link them later.
```
Still not a fan of using 3 lines for comments when 1 would do

fhennig replied:
I can try and remember, would be nice to update the linter/formatter if possible with this.

corneliusroemer reviewed (approved):
I'm not sure I'm convinced it's good to use "non strict" in docs as opposed to _lax_, don't think this was discussed much but not that important so I'll let go

fhennig commented:
Sorry about that! I had it in mind in the beginning but then lost track of it.

Thanks for approving 🙏

theosanderson commented:
> I'm not sure I'm convinced it's good to use "non strict" in docs as opposed to lax, don't think this was discussed much but not that important so I'll let go

I support the current state in the sense that the UI doesn't show the word "lax" so having it in the tooltip would IMO be confusing about what it would refer to

fhennig commented:
Yeah, initially I tried using a toggle with _strict_ and _lax_ on either side of it, but the toggle component was buggy so I dropped it. I linked to the bug report somewhere, it was a conflict somehow with the tailwind forms.

I decided it wasn't critical and moved on.
